Add articles from ada-howto repo (#623)
AdaCore/learn: content/courses/ada_2022_whats_new/chapters/image_attribute.rst

.. include:: ../../global.txt

:ada:`'Image` attribute for any type
====================================

.. note::

   Attribute :ada:`'Image` for any type is supported by

    * GNAT Community Edition 2020 and latter
    * GCC 11

:ada:`'Image` attribute for a value
-----------------------------------

> If you missed `Technical Corrigendum 1`_ changes (they were published
> in February 2016) then you probably don't know that the :ada:`'Image`
> attribute can be applied now to a value. So, instead of

Reviewer: Since our intent here is just to teach, I'd rather avoid something like "If you missed..." Instead, you could write this for example:

```suggestion
With the publication of the `Technical Corrigendum 1`_ in February
2016, the :ada:`'Image`attribute can now be applied to a value. So, 
instead of
```

.. code-block:: ada

   Ada.Text_IO.Put_Line (Ada.Text_IO.Page_Length'Image);

   Ada.Text_IO.Put_Line
     (Ada.Text_IO.Count'Image (Ada.Text_IO.Page_Length));

.. _`Technical Corrigendum 1`: https://reznikmm.github.io/ada-auth/rm-4-NC/RM-0-1.html
.. _name: https://reznikmm.github.io/ada-auth/rm-4-NC/RM-4-1.html#S0091

:ada:`'Image` attribute for any type
------------------------------------

Now you can apply :ada:`'Image` attribute for any type, including records,
array, access and private types. Let's see how this works. Define an array,
record, access types and corresponding objects:

.. code-block:: ada

   type Vector is array (Positive range <>) of Integer;

   V1 : aliased Vector := (1, 2, 3);

   type Text_Position is record
      Line, Column : Positive;
   end record;

   Pos : constant Text_Position := (Line => 10, Column => 3);

   type Vector_Access is access all Vector;

   V1_Ptr : constant Vector_Access := V1'Access;

Now you can convert these objects to string and print:

.. code-block:: ada

   Ada.Text_IO.Put_Line (V1'Image);

   Ada.Text_IO.Put_Line (Pos'Image);
   Ada.Text_IO.New_Line;
   Ada.Text_IO.Put_Line (V1_Ptr'Image);

> Output is:

Reviewer: Since the output is available when running the code snippet below, we don't really need to add it here.

> .. code-block:: ada

Reviewer: There's a Sphinx warning here because you're indicating that this block has Ada code when it's actually just the output of the application.
